implementations/rust/ockam/ockam_transport_websocket/src/common/router.rs

```diff
@@ -0,0 +1,217 @@
+            // Connection already exists
+            next = n;
+        } else {
+            // No existing connection
+            let peer_str;
+            if let Ok(s) = String::from_utf8(onward.deref().clone()) {
+                peer_str = s;
+            } else {
+                return Err(TransportError::UnknownRoute.into());
+            }
+
+            // TODO: Check if this is the hostname and we have existing/pending connection to this IP
+
+            let peer_addr_str = format!("{}#{}", self.addr_id, &peer_str);
+            if self.pending_connections.contains(&peer_addr_str) {
+                // We already trying to connect to this address - Requeue the message
+                ctx.forward(msg).await?;
+            } else if self.allow_auto_connection {
+                // Create connection
+                self.pending_connections.push(peer_addr_str);
+                let handle = self.new_handler(ctx).await?;
+                let _ = handle.connect(peer_str).await?;
+                // Requeue the message
+                ctx.forward(msg).await?;
+            } else {
+                return Err(TransportError::UnknownRoute.into());
+            }
+
+            return Ok(());
+        }
+
+        let _ = msg.transport_mut().onward_route.step()?;
+        // Modify the transport message route
+        msg.transport_mut()
+            .onward_route
+            .modify()
+            .prepend(next.clone());
+
+        // Send the transport message to the connection worker
+        ctx.send(next.clone(), msg).await?;
+
+        Ok(())
+    }
+}
+
+#[worker]
+impl Worker for Router {
+    type Message = RouterMessage;
+    type Context = Context;
+
+    async fn shutdown(&mut self, ctx: &mut Self::Context) -> Result<()> {
+        for (_accepts, self_addr) in self.map.iter() {
+            ctx.stop_worker(self_addr.clone()).await?;
+        }
+        Ok(())
+    }
+
+    async fn initialize(&mut self, ctx: &mut Context) -> Result<()> {
+        trace!("Initializing Router worker [addr_id = {}]", self.addr_id);
+        ctx.register(self.addr_id, ctx.address()).await?;
+        Ok(())
+    }
+
+    async fn handle_message(
+        &mut self,
+        ctx: &mut Context,
+        msg: Routed<RouterMessage>,
+    ) -> Result<()> {
+        let msg = msg.body();
+        use RouterMessage::*;
+        match msg {
+            Route(msg) => {
+                self.handle_route(ctx, msg).await?;
+            }
+            Register { accepts, self_addr } => {
+                self.handle_register(accepts, self_addr).await?;
+            }
+        };
+        Ok(())
+    }
+}
+
+pub struct RouterHandler {
+    ctx: Context,
+    addr: Address,
+}
+
+impl RouterHandler {
+    fn new(ctx: Context, addr: Address) -> Result<Self> {
+        debug!("Initializing RouterHandler instance");
+        Ok(Self { ctx, addr })
+    }
+
+    pub async fn shutdown(&self) -> Result<()> {
+        self.ctx.stop_worker(self.addr.clone()).await
```

Comment:
There may be many RouterHandler instances that use the same Router. We definitely don't want to stop Router when one of the Handlers stops.